src/server/pfs/server/driver.go

```diff
@@ -4482,14 +4482,35 @@ func (s *putFileServer) Peek() (*pfs.PutFileRequest, error) {
 	return req, nil
 }
 
-func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) (oneOff bool, repo string, branch string, err error) {
+func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) (oneOff bool, repo string, branch string, retErr error) {
+
 	var pr *io.PipeReader
 	var pw *io.PipeWriter
 	var req *pfs.PutFileRequest
 	var eg errgroup.Group
 	var rawCommitID string
 	var commitID string
 
+	// Always make sure our callbacks finish
+	defer func() {
+		err := eg.Wait()
+		if retErr == nil {
+			retErr = err
+		}
+	}()
+
+	// Always make sure we've closed any hanging pipes
+	defer func() {
+		if pw != nil {
+			if retErr != nil {
```

Comment:
Defer functions get called in LIFO, so it's possible that retErr gets set by the deferred function above but that means it doesn't get passed to `CloseWithError` because it's not set yet. I'm not sure if that's desired behavior or not. Does it make sense to combine these 2 deferred functions into a single one? Even if the semantics here are correct, I think it's probably clearer as one function, since it saves having to think about defer order.

Reply:
Sure, I'll merge them - I was counting on it being LIFO, but I can see how having them split would be confusing.